### PR TITLE
fix: delegate total and error member checks to WebIDL

### LINKED ISSUE #302: Why does PaymentRequest constructor throw on details.error
I'm confused as to why this throws in the constructor:

> If `details` contains a value for `error`, then throw a TypeError.

Why can't it be ignored? Or, why not make an DetailsUpdate dictionary type explicitly for this instead (i.e., with `updateWith()`)?

### PR DESCRIPTION
**Pull request by marcoscaceres**
* closes #302, #320

***
[Preview](https://s3.amazonaws.com/pr-preview/w3c/browser-payment-api/split-dictionary.html) | [Diff](https://s3.amazonaws.com/pr-preview/w3c/browser-payment-api/341ceab...840e7b8.html)

### CODE REVIEW

**Review by domenic — CHANGES_REQUESTED** [review]
This doesn't quite work as anticipated. In particular, the fact that the constructor requires a total is no longer enforced, and the updateWith() algorithm doesn't make sense, since it is looking for a total member, but PaymentDetailsUpdate does not contain a total member.

I think what you end up wanting is:

- PaymentDetailsBase: displayItems, shippingOptions, modifiers, total
- PaymentDetailsInit: PaymentDetailsBase + required total
- PaymentDetailsUpdate: PaymentDetailsBase + error

**Comment by marcoscaceres** [comment]
Ah yeah, of course. Will update.

File: index.html
Comment (marcoscaceres) [inline index.html]: Filed a bug with HTML tidy about the whitespace.

File: index.html
Comment (domenic) [inline index.html]: PaymentDetailsUpdate, right?

File: index.html
Comment (domenic) [inline index.html]: PaymentDetailsUpdate

File: index.html
Comment (domenic) [inline index.html]: PaymentDetailsBase.total, here and below

File: index.html
Comment (domenic) [inline index.html]: DictionaryName.memberName notation is not great and should not be exposed publicly. I'd do `<a>PaymentDetailsBase</a>'s <a data-lt="PaymentDetailsBase.total">total</a>`.

File: index.html
Comment (domenic) [inline index.html:1032]: It might be nice to nest these three sections under a single section (e.g. "Payment details dictionaries")

File: index.html
Comment (domenic) [inline index.html:1059]: PaymentDetailsBase

File: index.html
Comment (domenic) [inline index.html]: PaymentDetailsBase

**Review by domenic — COMMENTED** [review]
A number of nits, but the structure looks great now.

File: index.html
Comment (marcoscaceres) [reply index.html:1032]: Agree. I'll fix my find/replace screw ups first and do that as a new PR after.

**Comment by marcoscaceres** [comment]
Hopefully this is now good to go too.

**Comment by domenic** [comment]
Great, will check tomorrow, I really should stop working for the day :)

File: index.html
Comment (bentian) [inline index.html:1120]: Gecko hits [name duplication error](http://searchfox.org/mozilla-central/source/dom/bindings/parser/WebIDL.py#1867) while parsing member `total`. Not sure whether to revise WebIDL or adapt gecko to it.

Comment (bentian) [reply index.html:1120]: Seems WebIDL should be revised, per [section Dictionaries, WebIDL level 1](https://www.w3.org/TR/WebIDL-1/#idl-dictionaries):
> The identifier of a dictionary member MUST NOT be the same as that of another dictionary member defined on the dictionary or on that dictionary’s inherited dictionaries.

Comment (marcoscaceres) [reply index.html:1120]: (╯°□°）╯︵ ┻━┻)

Comment (domenic) [reply index.html:1120]: Filed https://github.com/heycam/webidl/issues/321. I guess the alternative is to just duplicate the contents into each dictionary instead of using inheritance... sigh.

**Comment by domenic** [comment]
OK, I guess the correct thing to do here is

- PaymentDetailsBase: displayItems, shippingOptions, modifiers
- PaymentDetailsInit: PaymentDetailsBase + required total
- PaymentDetailsUpdate: PaymentDetailsBase + error + non-required total

File: index.html
Comment (marcoscaceres) [reply index.html:1120]: I quite like Bz's proposal, as it avoids the duplication.

File: index.html
Comment (marcoscaceres) [reply index.html:1032]: This will be https://github.com/w3c/browser-payment-api/issues/424

**Comment by marcoscaceres** [comment]
Don't review this yet... still making changes. I pushed too early.

**Comment by marcoscaceres** [comment]
Hopefully ok now...

File: index.html
Comment (marcoscaceres) [reply index.html:1059]: this comment doesn't apply anymore, I think.

**Comment by bentian** [comment]
@marcoscaceres The latest IDL changes are good. No error while parsing on gecko.

**Comment by marcoscaceres** [comment]
@bentian, thanks for confirming. Just need ok from @domenic before merging.

**Comment by domenic** [comment]
I don't understand the point of PaymentDetailsMaybeMembers. Why not just inline `PaymentItem total` into `PaymentDetailsUpdate`, as I suggested in https://github.com/w3c/browser-payment-api/pull/425#issuecomment-282096201 ?

**Comment by marcoscaceres** [comment]
I thought having a generic place to dump optional things might be good for extensibility in the future. Don't mind switching tho.

**Comment by domenic** [comment]
Yeah, I think that would be significantly clearer :)

**Comment by marcoscaceres** [comment]
ok, now hopefully reflects https://github.com/w3c/browser-payment-api/pull/425#issuecomment-282096201 correctly 🤞.

File: index.html
Comment (domenic) [inline index.html:331]: ReSpec linking error introduced here

File: index.html
Comment (domenic) [inline index.html:1093]: A few lines above this there's a `<a>shippingOption</a>` that should be `<a data-lt="PaymentRequest.shippingOption">shippingOption</a>` I think. Preexisting problem maybe?

**Review by domenic — APPROVED** [review]
LGTM with nits

File: index.html
Comment (marcoscaceres) [reply index.html:331]: Yep, have a patch for ReSpec to start linking constructors. Will land hopefully in a week or so.

File: index.html
Comment (marcoscaceres) [reply index.html:1093]: Yeah, it's preexisting.

**Pull request merged**
